packages/interactor/src/interactor.ts

```diff
@@ -0,0 +1,86 @@
+import { converge } from './converge';
+import { InteractorSpecification } from './specification';
+import { Locator } from './locator';
+import { defaultOptions } from './options';
+import { NoSuchElementError, AmbigousElementError, NotAbsentError } from './errors';
+import { interaction, Interaction } from './interaction';
+
+export class Interactor<E extends Element> {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  private ancestors: Interactor<any>[] = [];
+
+  constructor(
+    public name: string,
+    private specification: InteractorSpecification<E>,
+    private locator: Locator<E>
+  ) {}
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  find<T extends Interactor<any>>(interactor: T): T {
+    return Object.create(interactor, {
+      ancestors: {
+        value: [...this.ancestors, this, ...interactor.ancestors]
+      }
+    });
+  }
+
+  get description(): string {
+    return this.ancestors.slice().reverse().reduce((desc, interactor) => {
+      return `${desc} within ${interactor.name} ${interactor.locator.description}`
+    }, `${this.name} ${this.locator.description}`);
+  }
+
+  private unsafeSyncResolve(): E {
+    let root = defaultOptions.document?.documentElement;
+
+    if(!root) {
+      throw new Error('must specify document');
+    }
+
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    let ancestorChain: Interactor<any>[] = [...this.ancestors, this];
```

Comment:
You may be able to remove the type annotation here too.

```suggestion
    let ancestorChain: Array<Interactor<any>> = [...this.ancestors, this];
```

Reply:
I'm not really sure why, but TypeScript complains when I remove this type annotation.

```
src/interactor.ts:44:75 - error TS2339: Property 'specification' does not exist on type 'this | Interactor<any>'.

44       let elements = Array.from(parentElement.querySelectorAll(interactor.specification.selector));
                                                                             ~~~~~~~~~~~~~
src/interactor.ts:45:70 - error TS2339: Property 'locator' does not exist on type 'this | Interactor<any>'.

45       let matchingElements = elements.filter((element) => interactor.locator.matches(element));
```

My guess is that this is also some obscure variance/subtyping issue.